sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceFactory.java

```diff
@@ -75,15 +76,15 @@ static BigtableServiceEntry create(ConfigId configId, BigtableService service) {
 
     @Override
     public void close() {
-      int refCount =
-          refCounts.getOrDefault(getConfigId().id(), new AtomicInteger(0)).decrementAndGet();
-      if (refCount < 0) {
-        LOG.error(
-            "close() Ref count is < 0, configId=" + getConfigId().id() + " refCount=" + refCount);
-      }
-      LOG.debug("close() for config id " + getConfigId().id() + ", ref count is " + refCount);
-      if (refCount == 0) {
-        synchronized (lock) {
+      synchronized (lock) {
+        int refCount =
+            refCounts.getOrDefault(getConfigId().id(), new AtomicInteger(0)).decrementAndGet();
+        if (refCount < 0) {
+          LOG.error(
+              "close() Ref count is < 0, configId=" + getConfigId().id() + " refCount=" + refCount);
+        }
+        LOG.debug("close() for config id " + getConfigId().id() + ", ref count is " + refCount);
+        if (refCount == 0) {
```

Comment:
decrementAndCount is thread safe, so I wonder what's the actual underlying issue here.

I also understand why < 0 is logging an error, but shouldn't we `.remove()` in that case too? Instead of doing `== 0` in the next if

Reply:
I think the underlying issue is the race between:
 - close() decrementing the counter and
 - getServiceForReading() incrementing the counter
Since decrementing the counter is outside of the lock, and we're getting the refCount number and do something with it, there could be a race. 

When `<0`, that means refCounts doesn't has a reference to configId, so we're not triggering the remove there.